core/src/main/java/org/apache/iceberg/avro/Avro.java

```diff
@@ -156,11 +170,175 @@ private CodecFactory codec() {
+      return this;
+    }
+
+    public DeleteWriteBuilder withKeyMetadata(EncryptionKeyMetadata metadata) {
+      this.keyMetadata = metadata;
+      return this;
+    }
+
+    public DeleteWriteBuilder equalityFieldIds(List<Integer> fieldIds) {
+      this.equalityFieldIds = fieldIds.stream().mapToInt(id -> id).toArray();
+      return this;
+    }
+
+    public DeleteWriteBuilder equalityFieldIds(int... fieldIds) {
+      this.equalityFieldIds = fieldIds;
+      return this;
+    }
+
+    public <T> EqualityDeleteWriter<T> buildEqualityWriter() throws IOException {
+      Preconditions.checkState(equalityFieldIds != null, "Cannot create equality delete file without delete field ids");
+
+      set("delete-type", "equality");
+      set("delete-field-ids", IntStream.of(equalityFieldIds)
+          .mapToObj(Objects::toString)
+          .collect(Collectors.joining(", ")));
+
+      // the appender uses the row schema without extra columns
+      appenderBuilder.schema(rowSchema);
+      if (createWriterFunc != null) {
+        appenderBuilder.createWriterFunc(createWriterFunc);
+      }
+
+      return new EqualityDeleteWriter<>(
+          appenderBuilder.build(), FileFormat.AVRO, location, spec, partition, keyMetadata, equalityFieldIds);
+    }
+
+    public <T> PositionDeleteWriter<T> buildPositionWriter() throws IOException {
+      Preconditions.checkState(equalityFieldIds == null, "Cannot create position delete file using delete field ids");
+
+      set("delete-type", "position");
+
+      // the appender uses the row schema wrapped with position fields
+      appenderBuilder.schema(new org.apache.iceberg.Schema(
+          MetadataColumns.DELETE_FILE_PATH,
+          MetadataColumns.DELETE_FILE_POS,
+          NestedField.optional(
+              MetadataColumns.DELETE_FILE_ROW_FIELD_ID, "row", rowSchema.asStruct(),
+              MetadataColumns.DELETE_FILE_ROW_DOC)));
+
+      if (createWriterFunc != null) {
+        appenderBuilder.createWriterFunc(
+            avroSchema -> new PositionDatumWriter<>(createWriterFunc.apply(avroSchema)));
+      }
+
+      return new PositionDeleteWriter<>(
+          appenderBuilder.build(), FileFormat.AVRO, location, spec, partition, keyMetadata, rowSchema.asStruct());
+    }
+  }
+
+  /**
+   * A {@link DatumWriter} implementation that wraps another to produce position delete rows.
+   *
+   * @param <D> the type of datum written as a deleted row
+   */
+  private static class PositionDatumWriter<D> implements DatumWriter<PositionDelete<D>> {
+    private static final ValueWriter<Object> PATH_WRITER = ValueWriters.strings();
+    private static final ValueWriter<Long> POS_WRITER = ValueWriters.longs();
+    private final DatumWriter<D> rowWriter;
+
+    private PositionDatumWriter(DatumWriter<D> rowWriter) {
+      this.rowWriter = rowWriter;
+    }
+
+    @Override
+    public void setSchema(Schema schema) {
+      rowWriter.setSchema(schema.getField("row").schema());
```

Comment:
The `row` field is an optional field,  it's possible that  `schema.getField("row")` would return a null.  We need to check nullability here I think.

Reply:
I misunderstood here,  although the row field is an optional field, `schema.getFiled("row")` won't be null and it will be the optional row field, please ignore this comment.